### ISSUE TITLE
Derving type fails for classes with a metaclass extending ABCMeta

### ISSUE THREAD

**Issue by cowfoodbagman** [microsoft/pyright#4925]
**Describe the bug**
When using a metaclass that inherits ABCMeta and implements `__call__` the derived type is not the class type. This is not the behavior when inheriting `type`. Additionally mypy is able to determine the correct type.

**Expected behavior**
The derived type is the type of the class

**Screenshots or Code**
```python
from abc import ABCMeta

class MyMeta(ABCMeta):
    def __call__(cls, *args, **kwds):
        return cls.__new__(cls, *args, **kwds)

class MyClass(metaclass=MyMeta):
    pass

instance = MyClass()

reveal_type(instance)
```

**Comment by erictraut** [comment]
This is related to [this issue](https://github.com/microsoft/pyright/issues/4911).

Mypy doesn't perform any type analysis for unannotated methods, nor does it ever infer function return types. 

You would need to supply type annotations to make this work correctly in pyright because the normal type inference rules don't work given the way that metaclasses work and the way that typeshed defines `type.__new__`.

Here's how this would look:

```python
T = TypeVar("T")

class MyMeta(ABCMeta):
    def __call__(cls: type[T], *args, **kwds) -> T:
        return cls.__new__(cls, *args, **kwds)
```

**Comment by cowfoodbagman** [comment]
In the proposed alternative generic classes don't keep their specific types, for example `List` instead of `List[str]`

**Comment by erictraut** [comment]
I'm not sure what you mean. The `List` class doesn't use a custom metaclass.

Are you talking about some user-defined generic class that uses this custom metaclass? Seems to work fine.

```python
from abc import ABCMeta
from typing import Generic, TypeVar

T = TypeVar("T")

class MyMeta(ABCMeta):
    def __call__(cls: type[T], *args, **kwds) -> T:
        return cls.__new__(cls, *args, **kwds)

class MyClass(Generic[T], metaclass=MyMeta):
    def __init__(self, x: T) -> None:
        ...

v1 = MyClass(0)
reveal_type(v1)  # MyClass[int]

v2 = MyClass("")
reveal_type(v2)  # MyClass[str]
```
